### PR TITLE
fix: don't re-import matchers for every match in top files

### PR DESCRIPTION
**Pull request by wbh1**
### What does this PR do?
Changes the `confirm_top` function to only import matchers once rather than for every possible match in a top file.

In large pillar environments with complex top files, I have observed that the most intensive (both measured by time and by CPU) operation is actually just the overhead of constantly re-importing all the matchers for each match in a top file.

By reusing work already done to avoid this gotcha elsewhere (see compound_match.py), Pillar render times and highstate times can be reduced by >6x, and load on the Salt master can be reduced by >5x.

### What issues does this PR fix or reference?
References/extends: #62295

### Previous Behavior
Matcher plugins are reloaded for every possible match in a top file (including multiple times for every part of a compound match).

### New Behavior
Matcher plugins are lazyloaded once and stored in the context dunder to be reused. 

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**

- [ ] Docs
- [x] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [ ] Tests written/updated

### Commits signed with GPG?
Yes

Please review [Salt's Contributing Guide](https://docs.saltproject.io/en/master/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

### CODE REVIEW

File: salt/matchers/confirm_top.py [hunk 1]
```diff
@@ -10,6 +10,14 @@
 log = logging.getLogger(__file__)
 
 
+def _load_matchers(opts):
+    """
+    Store matchers in __context__ so they're only loaded once
+    """
+    __context__["matchers"] = {}
```
Comment (cmcmarrow) [inline salt/matchers/confirm_top.py:17]: Can we remove `__context__["matchers"] = {}`

**Review by cmcmarrow — CHANGES_REQUESTED** [review]
Thanks for the PR, will need a change file https://docs.saltproject.io/en/master/topics/development/changelog.html

I'm putting in a PR to remove `__context__["matchers"] = {}` from the PR you linked

**Comment by cmcmarrow** [comment]
https://github.com/saltstack/salt/pull/64048

File: salt/matchers/confirm_top.py (hunk 1 above)
Comment (wbh1) [reply salt/matchers/confirm_top.py]: Done. Thanks!

**Review by cmcmarrow — APPROVED** [review]
Thanks @wbh1

**Review by dwoz — CHANGES_REQUESTED** [review]
I think this needs a test.

**Comment by wbh1** [comment]
> I think this needs a test.

@dwoz forgive my ignorance as I am not terribly familiar with the repository structure, but where in the tests directory would you propose that the test would go?

**Comment by dwoz** [comment]
> > I think this needs a test.
> 
> @dwoz forgive my ignorance as I am not terribly familiar with the repository structure, but where in the tests directory would you propose that the test would go?

They should go in one of:

`tests/pytests/{unit,functional,integration}/matchers`

**Comment by wbh1** [comment]
This is no longer needed.

@vzhestkov 's PR does the same thing and was merged: https://github.com/saltstack/salt/pull/66494

**Pull request closed without merging**